### ISSUE TITLE
PREPARE statement limit to 4KB

### ISSUE THREAD

**Issue by archongum** [prestosql/presto#628]
According to https://github.com/prestodb/presto/issues/5759
I think it should have a session configuration for this.

# reason for this
Because of prestosql capable of quering data cross different kinds of databases so long SQL is needed. Unfortunately, I have this kind of long SQLs....

**Comment by archongum** [comment]
On Presto UI, this statment executed successfully.
![1](https://user-images.githubusercontent.com/20929338/56008172-d752da00-5d0d-11e9-8986-f9f4932d0cff.jpg)

But return error
![2](https://user-images.githubusercontent.com/20929338/56008175-d8840700-5d0d-11e9-8d93-233a9c0882be.jpg)

```
SQL Error: Error executing query
  Error executing query
  Error executing query
    Error fetching next at http://datacenter3:18080/v1/statement/20190412_022832_01105_xz58u/x4c634b50e8534b90b9eb303de8dcdb85/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
    Error fetching next at http://datacenter3:18080/v1/statement/20190412_022832_01105_xz58u/x4c634b50e8534b90b9eb303de8dcdb85/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]

```

**Comment by findepi** [comment]
@archongum this is a known limitation. This is because prepared statements are sent in HTTP headers.
The plan is to sent them in POST body (there was some old PR doing this).
Current roadmap is to include this feature in "client protocol v2".

Prepared statements are sent in POST body when you're using Starburst Presto (with appropriate drivers),
this removes the limitation completely.

If you cannot use Starburst Presto, I recommend increasing HTTP headers limits (`http-server.max-request-header-size=64kB`).

**Comment by archongum** [comment]
> I recommend increasing HTTP headers limits (`http-server.max-request-header-size=64kB`).

By setting this session seems not working for me. It came out the same error.

**Comment by archongum** [comment]
@findepi Please reproduce it by executing this query.
[PREPARE statement1.txt](https://github.com/prestosql/presto/files/3072107/PREPARE.statement1.txt)

**Comment by ebyhr** [comment]
The query causes stackoverflow without the first line.

```sql
select 1 as id union 
select 1 as id union
...
```

```
 statement is too large (stack overflow during analysis)
```

**Comment by archongum** [comment]
@ebyhr It's just an example and I just want to simulate long SQL which I do have in prod. Try this query 
again please. 
[PREPARE-statement2.txt](https://github.com/prestosql/presto/files/3075612/PREPARE-statement2.txt)

**Comment by Jedda1314Jessie** [comment]
@ebyhr Actutally we want to execute a long statement whose size is 12 kb  in our production environment .What we sent to you is a simple to mock a statement of big size.

After we set up the http-server.max-request-header-size parameter to 64kb in config.properties, it still  could not been executed because it's size was more than 4kb.

**Comment by findepi** [comment]
@archongum @Jedda1314Jessie I assume you restarted Coordinator after setting `http-server.max-request-header-size` in config.properties.
When you tried again, what error did you get? Do you have full stacktrace?

Also, did you try Starburst Presto (with enterprise drivers)? Alternatively you can apply https://github.com/prestodb/presto/pull/5868 and build both Presto server and drivers.

**Comment by archongum** [comment]
# Env
config.properties
```bash
coordinator=true                        
node-scheduler.include-coordinator=true 
http-server.http.port=18080             
query.max-memory=2GB                    
query.max-memory-per-node=2GB           
query.max-total-memory-per-node=2GB     
discovery-server.enabled=true           
discovery.uri=http://datacenter4:18080                 
experimental.reserved-pool-enabled=false
http-server.max-request-header-size=64kB
```

# Step

## 1 execute **LONG** SQL(First execution)

### 1.1 mitmproxy(There is no `X-Presto-Prepared-Statement` parameter because it's a new connection)
![image](https://user-images.githubusercontent.com/20929338/56112669-52c2be80-5f8e-11e9-8e45-964eb1b803d3.png)
![image](https://user-images.githubusercontent.com/20929338/56112731-800f6c80-5f8e-11e9-989c-826b124ad239.png)


### 1.2 error message
```
SQL Error: Error executing query
  Error executing query
  Error executing query
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_065050_00071_abefk/xb84e529715774882a4398120d19ec533/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_065050_00071_abefk/xb84e529715774882a4398120d19ec533/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
```

## 2 execute **LONG** SQL again
### 2.1 mitmproxy(Also no `X-Presto-Prepared-Statement` parameter because the previous SQL failed)
![image](https://user-images.githubusercontent.com/20929338/56112886-e85e4e00-5f8e-11e9-81e1-fde1dbdfcd0b.png)
![image](https://user-images.githubusercontent.com/20929338/56112894-f0b68900-5f8e-11e9-932d-b45241b3f551.png)

### 2.2 error message
```
SQL Error: Error executing query
  Error executing query
  Error executing query
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_065744_00072_abefk/xdcdbe61ae2ac4dd89f62eaea66ec2b32/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_065744_00072_abefk/xdcdbe61ae2ac4dd89f62eaea66ec2b32/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
```

## 3 execute **SHORT** SQL
### 3.1 mitmproxy(no `X-Presto-Prepared-Statement`, previous SQL failed)
![image](https://user-images.githubusercontent.com/20929338/56112980-1d6aa080-5f8f-11e9-9e3b-23b74b8f9ceb.png)
![image](https://user-images.githubusercontent.com/20929338/56112991-22c7eb00-5f8f-11e9-8529-eb9df9a8e64d.png)


## 4 execute **SHORT** SQL again
### 4.1 mitmproxy(has `X-Presto-Prepared-Statement`, previous SQL succeed)
![image](https://user-images.githubusercontent.com/20929338/56113040-49862180-5f8f-11e9-9635-eae85c09519e.png)
![image](https://user-images.githubusercontent.com/20929338/56113050-4f7c0280-5f8f-11e9-8995-f9b875ef97ca.png)


## 5 execute **LONG** SQL
### 5.1 mitmproxy(has `X-Presto-Prepared-Statement`, previous SQL succeed)
![image](https://user-images.githubusercontent.com/20929338/56113098-70445800-5f8f-11e9-92bd-9deef24421a0.png)
![image](https://user-images.githubusercontent.com/20929338/56113107-75090c00-5f8f-11e9-8b5c-8cd6bb2630f8.png)


### 5.2 error message
```
SQL Error: Error executing query
  Error executing query
  Error executing query
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_070151_00075_abefk/x08640017eab44e0a9a0d22193c671cf4/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_070151_00075_abefk/x08640017eab44e0a9a0d22193c671cf4/2 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
```

## 6 execute **LONG** SQL
### 6.1 mitmproxy(no `X-Presto-Prepared-Statement` of **LONG** SQL, previous SQL failed)
![image](https://user-images.githubusercontent.com/20929338/56113153-9b2eac00-5f8f-11e9-8b6b-2a19de6d1ac3.png)
![image](https://user-images.githubusercontent.com/20929338/56113166-a1248d00-5f8f-11e9-8a0e-211cdb652517.png)

### 6.2 error message
```
SQL Error: Error executing query
  Error executing query
  Error executing query
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_070303_00076_abefk/x8835a3f742474146bc02cbb2949a6492/1 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
    Error fetching next at http://datacenter4:18080/v1/statement/20190415_070303_00076_abefk/x8835a3f742474146bc02cbb2949a6492/1 returned an invalid response: JsonResponse{statusCode=500, statusMessage=Server Error, headers={connection=[close]}, hasValue=false} [Error: ]
```

**Comment by archongum** [comment]
### Env
* presto: StarBurst Presto 302-e.7
* presto-jdbc: StarBurst Presto jdbc 302-e.7
* NiFi: 1.8.0

@findepi It seems that StarBurst Presto not support Connection Pool?
![image](https://user-images.githubusercontent.com/20929338/56190815-ad771b80-605d-11e9-812b-01940a8b8691.png)

If I change jdbc to `presto-jdbc-308.jar` in NiFi (still StarBurst Presto Server) then it works fine.

**Comment by elonlo** [comment]
That is because the jetty response header size >= 8kB, you need add http-server.max-response-header-size=64kB to override default size
ref to [Airlift](https://github.com/airlift/airlift/pull/727)

**Comment by elonlo** [comment]
That is because the jetty response header size >= 8kB, you need also add http-server.max-response-header-size=64kB to override default size
ref to [airlift](https://github.com/airlift/airlift/pull/727)

**Comment by archongum** [comment]
@elonlo Thanks! Fixed by [airlift/airlift #727](https://github.com/airlift/airlift/pull/727)

add to `config.properties`
```bash
http-server.max-request-header-size=64kB
http-server.max-response-header-size=64kB
```

effective on `prestosql 310`
